### PR TITLE
add pyupgrade to pre-commit

### PR DESCRIPTION
**Pull request by basnijholt**
As a followup from https://github.com/QCoDeS/Qcodes/pull/1945.

This adds pyupgrade as pre-commit filter.

This PR is blocked by
- #2201
- #2202
- #2203
- #2204
- #2205
- #2206
- #2207
- #2208
- #2209
- #2210
- #2211

### CODE REVIEW

File: qcodes/instrument/base.py [hunk 1]
```diff
@@ -266,7 +266,7 @@ def print_readable_snapshot(self, update: bool = False,
                 # this may be a multi parameter
                 unit = snapshot['parameters'][par].get('units', None)
             if isinstance(val, floating_types):
-                msg += f'\t{val:.5g} '  # type: ignore[str-format]
+                msg += f'\t{val:.5g} '
```
Comment (jenshnielsen) [inline qcodes/instrument/base.py:269]: It's not unused but required when you check with the numpy stubs installed. It if was unrequied our CI would have flagged it given the current config.

Comment (jenshnielsen) [reply qcodes/instrument/base.py:269]: In any case it does not belong in this pr

File: qcodes/_version.py [hunk 2]
```diff
@@ -81,7 +80,7 @@ def run_command(commands, args, cwd=None, verbose=False, hide_stderr=False,
                                  stderr=(subprocess.PIPE if hide_stderr
                                          else None))
             break
-        except EnvironmentError:
+        except OSError:
```
Comment (jenshnielsen) [inline qcodes/_version.py:83]: This file is vendored from versioneer so it should not be modified

File: qcodes/instrument/base.py (hunk 1 above)
Comment (basnijholt) [reply qcodes/instrument/base.py:269]: I based that on the [CI failure of the previous commit](https://dev.azure.com/ms-quantum-public/Microsoft%20Quantum%20(public)/_build/results?buildId=15137&view=logs&jobId=347123c7-c6ed-5812-0c14-06a064703b24&j=347123c7-c6ed-5812-0c14-06a064703b24&t=5f8e1bf4-3899-52e0-b0b2-57867ecb13e6):
```
Starting: mypy
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.171.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
========================== Starting Command Output ===========================
/bin/bash --noprofile --norc /home/vsts/work/_temp/41993469-0857-4a4e-9ebb-af6c368b7f56.sh
qcodes/instrument/base.py:269: error: unused 'type: ignore' comment
Found 1 error in 1 file (checked 395 source files)

##[error]Bash exited with code '1'.
Finishing: mypy
```

File: versioneer.py [hunk 3]
```diff
@@ -276,7 +275,6 @@
 
 """
 
-from __future__ import print_function
```
Comment (jenshnielsen) [inline versioneer.py:279]: This file is vendored and should not be modified

**Comment by jenshnielsen** [comment]
While many of the changes are nice I am not sure I think this is the best approch. 

IMHO these changes need careful review and that can only really be done on small prs. 
and the way to go would be to modernize <10 files or so pr until there are nothing left. 
Then the hook can be activated. There also needs to be an option to skip vendored files which should never be modified

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -7,3 +7,9 @@ repos:
     - id: check-yaml
     - id: debug-statements
     - id: check-ast
+  - repo: https://github.com/asottile/pyupgrade
+    rev: v2.7.2
+    hooks:
+    - id: pyupgrade
+      args: ['--py37-plus']
+      exclude: (^_version/|^versioneer/)
```
Comment (basnijholt) [inline .pre-commit-config.yaml:15]: The ignore pattern doesn't seem to work.

src: https://github.com/psf/black/issues/395

**Review by jenshnielsen — APPROVED** [review]

**Comment by jenshnielsen** [comment]
@astafan8 do you approve ?

**Review by astafan8 — APPROVED** [review]

**Pull request merged**